tutorials/week4-react.md

```diff
@@ -0,0 +1,511 @@
+repeatable html element with built-in state, business logic, and a
+lifecycle. The component may be something as simple as a single html
+element such as an input box, or a button, or a complex entity made up
+of other components.
+
+Components are the basic building blocks of a React application and they allow the developer to split the UI into independent and reusable
+pieces where each piece can be used in isolation. 
+### Creating A Function Component in React
+
+<a href="https://codesandbox.io/s/nervous-morse-o3pwqm?file=/src/tutorial/CreateFunctionComponent.tsx" target="_blank">view in sandbox</a>
+
+The simplest method to define a component is to write a function in Javascript. These components are also widely referred as functional stateless components but in the recent versions of React, they have the capabilities to contain state as well.
+```ts
+import * as React from "react";
+
+interface Props = {name?: string};
+
+function App (props: Props) {
+  return (
+      <div className="App">
+        <h1>Welcome to React with Typescript Tutorial.</h1>
+  </div>
+);
+}
+
+export default App;
+```
+
+### A Few Things to Note About React Components:
+
+- The root (App) component is the entry point for the React App and all other components are nested in it.
+- We define a function component using a javascript function, passing the props type in as a generic parameter.
+- The import statement is used to import the public classes/functions from the `react` library.
+- A function can return a single top level element. 
+  - div is the top level element in this case and other elements can be nested in it.
+    - The attribute `className` is used to specify a CSS class name if CSS properties have been defined seperately for a class.
+    - `className` attribute is used to set the value of an element’s class attribute. Using this property, the user can set the class of an element to the desired class.
+- The round brackets (()) after return are used to span a JSX/TSX element across multiple lines.
+- At last , the component needs to be exported from the current file, so that it can be imported somewhere else and can be used either in isolation or combination with other components for rendering on the UI.
+- Elements on one line can be returned directly.
+- Each instance of a component creates a new element independent of other instances of the component.
+- Each component has it's own state, props, and lifecycle (which will be explored later in the tutorial).
+
+## Template For Structure of Function Component
+
+<a href="https://codesandbox.io/s/nervous-morse-o3pwqm?file=/src/tutorial/TemplateForFunction.tsx" target="_blank">view in sandbox</a>
+
+```ts
+import * as React from "react";
+
+interface Props = {property1?: value1, property2?: value2}
+
+function ComponentName (props: Props) {
+ const handler = () => {
+ ...
+ };
+
+ return (
+  <div>Our TSX</div>
+ );
+};
+ComponentName.defaultProps = {
+ ...
+};
+
+export default ComponentName;
+```
```

Comment:
Also, Can we define types of value1 and value2 ? Some students might try to copy this snippet and practice it. Also, can we use comments instead of "..." ?

here's my version:

```
import * as React from "react";

type value1 = string;
type value2 = boolean;

interface Props {
  property1?: value1;
  property2?: value2;
}

const ComponentName = (props: Props) => {
  const handler = () => {
    //  code
  };

  return <div>Our TSX</div>;
};

ComponentName.defaultProps = {
  // code
};

export default ComponentName;

```